torch/_functorch/_aot_autograd/fx_passes.py

```diff
@@ -0,0 +1,152 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import torch
+from torch import fx
+
+
+def is_primal(node: fx.Node) -> bool:
+    return node.op == "placeholder" and "tangents" not in str(node.target)
+
+
+def move_resize_zero_to_end_of_graph(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    for n in node_list:
+        if (
+            n.op == "call_function"
+            and n.target is torch.ops.inductor.resize_storage_bytes_.default
+            and n.args[1] == 0
+        ):
+            resize_node = n
+            with graph.inserting_before(return_node):
+                new_resize_node = graph.call_function(
+                    torch.ops.inductor.resize_storage_bytes_.default, resize_node.args
+                )
+            graph.erase_node(resize_node)
+
+
+def refunctionalize_set(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    primal_inputs = [*filter(is_primal, node_list)]
+    primal_input_to_set_idx = defaultdict(list)
+    node_to_idx = {n: i for i, n in enumerate(node_list)}
+    set_nodes_to_be_deleted = set()
+
+    # Step 1: Clean up to enforce `.set_(primal_X, ...)` usage.
+    # We replace:
+    # ```
+    # set_1 = aten.set_(primal_X, Y1)
+    # set_2 = aten.set_(set_1, Y2)
+    # ```
+    # to:
+    # ```
+    # set_1 = aten.set_(primal_X, Y1)
+    # set_2 = aten.set_(primal_X, Y2)
+    # ```
+    for i, n in enumerate(node_list):
+        if (
+            n.op == "call_function"
+            and n.target is torch.ops.aten.set_.source_Tensor
+            and n.args[0].target is torch.ops.aten.set_.source_Tensor
+            and n.args[0].args[0] in primal_inputs
+        ):
+            n.args = (n.args[0].args[0], n.args[1])
+
+    # Step 2: Re-functionalizing `.set_(primal_X, ...)`.
+    # Replacement conditions:
+    # 1. `.set_(primal_X, Y)` exists in graph.
+    # 2. `.set_(Y, ...)` is not called between primal_X's this `.set_(primal_X, Y)` and its next `.set_(primal_X, ...)` (or end of graph)  # noqa: B950
```

Comment:
I think my two main questions are:

(1) how does FSDP end up in a situation where we see `.set_(Y, ...)` in the graph (where Y is not a primal)?

(2) I think the main thing I was expecting to see in the graph pass is handling for the set_() + aliasing situation:
```
def f(param: [4], param_aliased: [2, 2]):
    param.set_(ag_buffer)
    return torch.matmul(param_aliased, ...)
```

Do we still need to handle that?

Reply:
(1) this never happens, so I should just ban it

Reply:
Graphs before this PR: https://gist.github.com/yf225/5f056bf14312241940d382a63e716ba5

Reply:
To recap, how we got to this PR:
1. We can't functionalize `.set_` and `.resize_(0)` (to avoid subclass alias deduplication problem), thus those two mutation ops are in the middle of the AOT joint graph.
2. We want to move those two mutation ops to the end of the AOT joint graph. Easy for `.resize_(0)`, but for `.set_` we need this `refunctionalize_set()` pass.
3. After refunctionalizing `.set_`, it only appears at the end of AOT joint graph now, which is good.
4. However, Traceable FSDP2 has another constraint that the FWD primal that's `.set_` into must be saved in FWD graph output. But after refunctionalizing `.set_`, by default the partitioner saves an alias of primal instead of the primal itself.
5. Hence we need the special pass in partitioner to save the primal instead of that primal alias.

Reply:
> (1) how does FSDP end up in a situation where we see .set_(Y, ...) in the graph (where Y is not a primal)?

Added assertion to throw error in this case

Reply:
> (1) how does FSDP end up in a situation where we see .set_(Y, ...) in the graph (where Y is not a primal)?

Also discovered one nuance: for "`.set_(primal_X, primal_Y)` then `.set_(primal_Y, ...)`" case (`test_tensor_set_data` in test_repros.py does this), we need to not throw error, but we still can't do this  functionalization pass (and will instead skip it) because the `.set_` ordering needs to be respected.

Reply:
hmmm... @yf225 IIRC, that test was added to explicitly support `.data=` usage in FSDP: https://github.com/pytorch/pytorch/pull/113080. I think the idea is that FSDP1 uses the `.data =  foo` pattern, and dynamo also has special logic to re-write this into `set_()`.

Which makes me wonder, do you think we need to support the "more general" usage of `set_` in FSDP1 then, as part of this functionalization pass? Or would it be reasonable to xfail that test and declare it unsupported (my understanding is that FSDP1 isn't "really" traceable today anyway?).

Reply:
yeah I feel that we should just error out for the FSDP1 use case for now. We can revisit the FSDP1 eager code later to see if we can bring it closer to FSDP2 impl.